File: mac/Sources/VibeBoardKit/Input/GestureRouter.swift
import Foundation

public struct GesturePolicy: Equatable, Sendable {
    public let doubleClickWindowMilliseconds: UInt32
    public let longPressThresholdMilliseconds: UInt32
    public let derivesDoubleClick: Bool
    public let derivesLongPress: Bool

    public init(
        doubleClickWindowMilliseconds: UInt32,
        longPressThresholdMilliseconds: UInt32,
        derivesDoubleClick: Bool,
        derivesLongPress: Bool
    ) throws {
        guard doubleClickWindowMilliseconds > 0 else {
            throw GestureError.invalidPolicy("doubleClickWindowMilliseconds")
        }
        guard longPressThresholdMilliseconds > 0 else {
            throw GestureError.invalidPolicy("longPressThresholdMilliseconds")
        }
        self.doubleClickWindowMilliseconds = doubleClickWindowMilliseconds
        self.longPressThresholdMilliseconds = longPressThresholdMilliseconds
        self.derivesDoubleClick = derivesDoubleClick
        self.derivesLongPress = derivesLongPress
    }
}

public enum DeviceKeyEvent: Equatable, Sendable {
    case down(CanonicalKey)
    case up(CanonicalKey, durationMilliseconds: UInt32?)
    case click(CanonicalKey, durationMilliseconds: UInt32?)
    case disconnect
}

public struct RoutedGesture: Equatable, Sendable {
    public let key: CanonicalKey
    public let gesture: KeyGesture

    public init(key: CanonicalKey, gesture: KeyGesture) {
        self.key = key
        self.gesture = gesture
    }
}

public enum GestureError: Error, Equatable, Sendable {
    case invalidPolicy(String)
    case duplicateDown(CanonicalKey)
    case upWithoutDown(CanonicalKey)
    case timestampRegression
}

public struct GestureRouter: Sendable {
    private struct PendingClick: Sendable {
        let timestampMilliseconds: UInt64
    }

    private let policy: GesturePolicy
    private var downTimestamps: [CanonicalKey: UInt64] = [:]
    private var pendingClicks: [CanonicalKey: PendingClick] = [:]
    private var suppressedClicks: Set<CanonicalKey> = []
    private var lastTimestampMilliseconds: UInt64?

    public init(policy: GesturePolicy) {
        self.policy = policy
    }

    public mutating func handle(_ event: DeviceKeyEvent, at timestampMilliseconds: UInt64) throws -> [RoutedGesture] {
        if case .disconnect = event {
            resetSession()
            return []
        }
        try validateTimestamp(timestampMilliseconds)
        var output = flushExpired(at: timestampMilliseconds)

        switch event {
        case let .down(key):
            guard downTimestamps[key] == nil else { throw GestureError.duplicateDown(key) }
            suppressedClicks.remove(key)
            downTimestamps[key] = timestampMilliseconds

        case let .up(key, _):
            guard let downTimestamp = downTimestamps.removeValue(forKey: key) else {
                throw GestureError.upWithoutDown(key)
            }
            if policy.derivesLongPress {
                let measured = timestampMilliseconds - downTimestamp
                if measured >= UInt64(policy.longPressThresholdMilliseconds) {
                    pendingClicks.removeValue(forKey: key)
                    suppressedClicks.insert(key)
                    output.append(RoutedGesture(key: key, gesture: .long))
                }
            }

        case let .click(key, _):
            if suppressedClicks.remove(key) != nil {
                break
            }
            if policy.derivesDoubleClick {
                if let pending = pendingClicks[key],
                   timestampMilliseconds - pending.timestampMilliseconds <= UInt64(policy.doubleClickWindowMilliseconds) {
                    pendingClicks.removeValue(forKey: key)
                    output.append(RoutedGesture(key: key, gesture: .double))
                } else {
                    pendingClicks[key] = PendingClick(timestampMilliseconds: timestampMilliseconds)
                }
            } else {
                output.append(RoutedGesture(key: key, gesture: .single))
            }

        case .disconnect:
            // Handled before timestamp validation so a new session can reset its clock.
            break
        }
        return output
    }

    public mutating func flush(at timestampMilliseconds: UInt64) throws -> [RoutedGesture] {
        try validateTimestamp(timestampMilliseconds)
        return flushExpired(at: timestampMilliseconds)
    }

    private mutating func resetSession() {
        downTimestamps.removeAll(keepingCapacity: true)
        pendingClicks.removeAll(keepingCapacity: true)
        suppressedClicks.removeAll(keepingCapacity: true)
        lastTimestampMilliseconds = nil
    }

    private mutating func validateTimestamp(_ timestampMilliseconds: UInt64) throws {
        if let lastTimestampMilliseconds, timestampMilliseconds < lastTimestampMilliseconds {
            throw GestureError.timestampRegression
        }
        lastTimestampMilliseconds = timestampMilliseconds
    }

    private mutating func flushExpired(at timestampMilliseconds: UInt64) -> [RoutedGesture] {
        let window = UInt64(policy.doubleClickWindowMilliseconds)
        let expired = pendingClicks.compactMap { key, pending -> (CanonicalKey, UInt64)? in
            guard timestampMilliseconds - pending.timestampMilliseconds > window else { return nil }
            return (key, pending.timestampMilliseconds)
        }.sorted {
            if $0.1 == $1.1 { return $0.0.rawValue < $1.0.rawValue }
            return $0.1 < $1.1
        }
        for (key, _) in expired {
            pendingClicks.removeValue(forKey: key)
        }
        return expired.map { RoutedGesture(key: $0.0, gesture: .single) }
    }
}

File: win/src/VibeBoardKit/Input/GestureRouter.cs
using System;
using System.Collections.Generic;

namespace VibeBoardKit.Input;

/// <summary>Reasons for which gesture routing fails.</summary>
public enum GestureErrorReason
{
    InvalidPolicy,
    DuplicateDown,
    UpWithoutDown,
    TimestampRegression
}

/// <summary>
/// Thrown when a <see cref="GestureRouter"/> receives an invalid event sequence or policy.
/// Mirrors the macOS <c>GestureError</c>.
/// </summary>
public sealed class GestureException : Exception
{
    public GestureErrorReason Reason { get; }
    public CanonicalKey? Key { get; }
    public string? Detail { get; }

    public GestureException(GestureErrorReason reason, CanonicalKey? key = null, string? detail = null)
        : base(BuildMessage(reason, key, detail))
    {
        Reason = reason;
        Key = key;
        Detail = detail;
    }

    public static GestureException InvalidPolicy(string field) =>
        new(GestureErrorReason.InvalidPolicy, detail: field);

    public static GestureException DuplicateDown(CanonicalKey key) =>
        new(GestureErrorReason.DuplicateDown, key);

    public static GestureException UpWithoutDown(CanonicalKey key) =>
        new(GestureErrorReason.UpWithoutDown, key);

    public static GestureException TimestampRegression() =>
        new(GestureErrorReason.TimestampRegression);

    private static string BuildMessage(GestureErrorReason reason, CanonicalKey? key, string? detail) =>
        reason switch
        {
            GestureErrorReason.InvalidPolicy => $"Invalid gesture policy: {detail}",
            GestureErrorReason.DuplicateDown => $"Duplicate down for key {key}",
            GestureErrorReason.UpWithoutDown => $"Up without down for key {key}",
            GestureErrorReason.TimestampRegression => "Timestamp regression",
            _ => reason.ToString()
        };
}

/// <summary>
/// Configures how a <see cref="GestureRouter"/> derives double-clicks and long-presses.
/// Time thresholds default to 350 ms (double-click window) and 500 ms (long-press threshold).
/// </summary>
public sealed class GesturePolicy
{
    /// <summary>The maximum interval between two clicks for them to count as a double click.</summary>
    public long DoubleClickWindowMilliseconds { get; }

    /// <summary>The minimum press duration for a press to count as a long press.</summary>
    public long LongPressThresholdMilliseconds { get; }

    /// <summary>Whether the router derives double-clicks from firmware click events.</summary>
    public bool DerivesDoubleClick { get; }

    /// <summary>Whether the router derives long-presses from down/up timing.</summary>
    public bool DerivesLongPress { get; }

    public GesturePolicy(
        long doubleClickWindowMilliseconds = 350,
        long longPressThresholdMilliseconds = 500,
        bool derivesDoubleClick = true,
        bool derivesLongPress = true)
    {
        if (doubleClickWindowMilliseconds <= 0)
            throw GestureException.InvalidPolicy("doubleClickWindowMilliseconds");
        if (longPressThresholdMilliseconds <= 0)
            throw GestureException.InvalidPolicy("longPressThresholdMilliseconds");

        DoubleClickWindowMilliseconds = doubleClickWindowMilliseconds;
        LongPressThresholdMilliseconds = longPressThresholdMilliseconds;
        DerivesDoubleClick = derivesDoubleClick;
        DerivesLongPress = derivesLongPress;
    }
}

/// <summary>
/// A raw key event from the device. A discriminated union mirroring the macOS
/// <c>DeviceKeyEvent</c>: a key-down, a key-up, a firmware click, or a disconnect.
/// </summary>
public abstract class DeviceKeyEvent
{
    private DeviceKeyEvent() { }

    /// <summary>The key was pressed down.</summary>
    public sealed class Down : DeviceKeyEvent
    {
        public CanonicalKey Key { get; }
        public Down(CanonicalKey key) => Key = key;
    }

    /// <summary>The key was released.</summary>
    public sealed class Up : DeviceKeyEvent
    {
        public CanonicalKey Key { get; }
        public long? DurationMilliseconds { get; }
        public Up(CanonicalKey key, long? durationMilliseconds = null)
        {
            Key = key;
            DurationMilliseconds = durationMilliseconds;
        }
    }

    /// <summary>A firmware click (the device already decided this was a click).</summary>
    public sealed class Click : DeviceKeyEvent
    {
        public CanonicalKey Key { get; }
        public long? DurationMilliseconds { get; }
        public Click(CanonicalKey key, long? durationMilliseconds = null)
        {
            Key = key;
            DurationMilliseconds = durationMilliseconds;
        }
    }

    /// <summary>The device disconnected; all pending state should be cleared.</summary>
    public sealed class Disconnect : DeviceKeyEvent { }
}

/// <summary>A gesture that the router has decided to emit for a key.</summary>
public readonly struct RoutedGesture : IEquatable<RoutedGesture>
{
    public CanonicalKey Key { get; }
    public KeyGesture Gesture { get; }

    public RoutedGesture(CanonicalKey key, KeyGesture gesture)
    {
        Key = key;
        Gesture = gesture;
    }

    public bool Equals(RoutedGesture other) => Key == other.Key && Gesture == other.Gesture;
    public override bool Equals(object? obj) => obj is RoutedGesture other && Equals(other);
    public override int GetHashCode() => HashCode.Combine(Key, Gesture);
    public static bool operator ==(RoutedGesture left, RoutedGesture right) => left.Equals(right);
    public static bool operator !=(RoutedGesture left, RoutedGesture right) => !left.Equals(right);
}

/// <summary>
/// Translates a stream of <see cref="DeviceKeyEvent"/>s into <see cref="RoutedGesture"/>s
/// using a <see cref="GesturePolicy"/>. This is a precise port of the macOS
/// <c>GestureRouter</c> struct.
/// </summary>
public sealed class GestureRouter
{
    private readonly GesturePolicy _policy;
    private readonly Dictionary<CanonicalKey, long> _downTimestamps = new();
    private readonly Dictionary<CanonicalKey, long> _pendingClicks = new();
    private readonly HashSet<CanonicalKey> _suppressedClicks = new();
    private long? _lastTimestamp;

    public GestureRouter(GesturePolicy policy)
    {
        _policy = policy ?? throw new ArgumentNullException(nameof(policy));
    }

    /// <summary>
    /// Routes a single device event, returning any gestures that are now determined.
    /// <paramref name="timestampMs"/> must be monotonically non-decreasing, except that
    /// a <see cref="DeviceKeyEvent.Disconnect"/> resets the clock and may carry an
    /// earlier timestamp.
    /// </summary>
    public List<RoutedGesture> Handle(DeviceKeyEvent @event, long timestampMs)
    {
        if (@event is null) throw new ArgumentNullException(nameof(@event));

        // Disconnect resets the session and bypasses timestamp validation so a new
        // session can start its clock from zero.
        if (@event is DeviceKeyEvent.Disconnect)
        {
            ResetSession();
            return new List<RoutedGesture>();
        }

        ValidateTimestamp(timestampMs);
        var output = FlushExpired(timestampMs);

        switch (@event)
        {
            case DeviceKeyEvent.Down down:
                if (_downTimestamps.ContainsKey(down.Key))
                    throw GestureException.DuplicateDown(down.Key);
                _suppressedClicks.Remove(down.Key);
                _downTimestamps[down.Key] = timestampMs;
                break;

            case DeviceKeyEvent.Up up:
                if (!_downTimestamps.Remove(up.Key, out var downTimestamp))
                    throw GestureException.UpWithoutDown(up.Key);
                if (_policy.DerivesLongPress)
                {
                    long measured = timestampMs - downTimestamp;
                    if (measured >= _policy.LongPressThresholdMilliseconds)
                    {
                        _pendingClicks.Remove(up.Key);
                        _suppressedClicks.Add(up.Key);
                        output.Add(new RoutedGesture(up.Key, KeyGesture.Long));
                    }
                }
                break;

            case DeviceKeyEvent.Click click:
                if (_suppressedClicks.Remove(click.Key))
                    break;
                if (_policy.DerivesDoubleClick)
                {
                    if (_pendingClicks.TryGetValue(click.Key, out var pending) &&
                        timestampMs - pending <= _policy.DoubleClickWindowMilliseconds)
                    {
                        _pendingClicks.Remove(click.Key);
                        output.Add(new RoutedGesture(click.Key, KeyGesture.Double));
                    }
                    else
                    {
                        _pendingClicks[click.Key] = timestampMs;
                    }
                }
                else
                {
                    output.Add(new RoutedGesture(click.Key, KeyGesture.Single));
                }
                break;
        }

        return output;
    }

    /// <summary>
    /// Emits single gestures for any pending clicks whose double-click window has expired.
    /// </summary>
    public List<RoutedGesture> Flush(long timestampMs)
    {
        ValidateTimestamp(timestampMs);
        return FlushExpired(timestampMs);
    }

    private void ResetSession()
    {
        _downTimestamps.Clear();
        _pendingClicks.Clear();
        _suppressedClicks.Clear();
        _lastTimestamp = null;
    }

    private void ValidateTimestamp(long timestampMs)
    {
        if (_lastTimestamp is { } last && timestampMs < last)
            throw GestureException.TimestampRegression();
        _lastTimestamp = timestampMs;
    }

    private List<RoutedGesture> FlushExpired(long timestampMs)
    {
        long window = _policy.DoubleClickWindowMilliseconds;

        var expired = new List<(CanonicalKey Key, long Timestamp)>();
        foreach (var kv in _pendingClicks)
        {
            if (timestampMs - kv.Value > window)
                expired.Add((kv.Key, kv.Value));
        }

        // Stable ordering: by pending timestamp, then by canonical key, matching the
        // macOS implementation's sort so emitted singles are deterministic.
        expired.Sort((a, b) =>
        {
            int c = a.Timestamp.CompareTo(b.Timestamp);
            if (c != 0) return c;
            return ((int)a.Key).CompareTo((int)b.Key);
        });

        foreach (var (key, _) in expired)
            _pendingClicks.Remove(key);

        var result = new List<RoutedGesture>(expired.Count);
        foreach (var (key, _) in expired)
            result.Add(new RoutedGesture(key, KeyGesture.Single));
        return result;
    }
}
